### PR TITLE
L2 Governance Hotfix

### PR DESCRIPTION
**Pull request by soloseng**
### Description

Implements 2 phase hotfixes (Approval & execute):

- The Security Council multisig must approve a hotfix for it to be executed.
- The Approver multisig must approve a hotfix for it to be executed.
- Anyone with the raw transaction that derives a hotfix hash can "execute" at anytime.

### Tested

Unit tested.

### Related issues

- Fixes https://github.com/celo-org/celo-blockchain-planning/issues/199

- Fixes https://github.com/celo-org/celo-blockchain-planning/issues/198

### Backwards compatibility

Maintains backwards compatibility with L1.

### CODE REVIEW

**Comment by soloseng** [comment]
Worked on implementing a hotfix mechanism using a Security Council multisig (number of signer TBD). This new process, however, is potentially less secure than the original validator hotfix process. 

The hotfix process went from a 3 step to a 2 step approval process, as the `prepare` step (using epoch number) has been deprecated.

The original validator hotfix process required a byzantine quorum of validators in order to execute a hotfix. Because validators had a financial incentive to not be malicious, this guaranteed that a set threshold of healthy, non-malicious validators approved the hotfix. In addition to that, since the list of validators approving the hotfix depended on current set of validators at the time of approval, it was much more difficult for validators to collude and approve a malicious hotfix.

With the `prepare` step being deprecated, this leaves a hotfix proposal open to being executed at anytime after it was approved. This is divergent from the original hotfix process, as with each new epoch, the `hotfixWhitelistValidatorTally` could change, requiring validators to reapprove a hotfix that was not executed in the allotted timeframe.

Using the multisig approach on L2, the list of signer would remain fixed. Given that there are no obvious incentives for Security Council signers to act for the greater good of the network, this increases the risk of collusion between Security Council signers, and the risk of a malicious hotfix being approved and executed. The main safety against malicious hotfix is based on the assumption that there will be no overlap or collusion between approvers and the Security Council signers.

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [inline packages/protocol/contracts/governance/Governance.sol]: Should add a way of invalidating the L2 hotfix if too much time has elapsed between approval and execution.

**Comment by soloseng** [comment]
> With the prepare step being deprecated, this leaves a hotfix proposal open to being executed at anytime after it was approved. This is divergent from the original hotfix process, as with each new epoch, the hotfixWhitelistValidatorTally could change, requiring validators to reapprove a hotfix that was not executed in the allotted timeframe.

I added a time limit for the prepareHotfix step on L2. This requires the approvers and Security Council to re-approve and prepare the hotfix once the `hotfixes[hash].executionTimeLimit` has elapsed.

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol]: prepareHotfix will reset the hotfix record if the `executionTimeLimit` has already been reached. This will require an additional round of approval by the approvers and Security Council, as well as an additional `prepareHotfix` step.

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol]: Can we change the message so it reflects the condition ?

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol:662]: This condition should be in root of this function. 

if (msg.sender == approver) shoudn't be considered in L2 anymore

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol]: We should keep both comments for L1 and L2

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol:664]: there should be check whether msg.sender == securityCouncil

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol]: What is the reasoning behind resetting hotfix ?

Also it should probably be new function

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol]: Since we are reading hotfixes[hash] we should extract it to local storage variable to save some gas

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol:729]: Can we use getHotfixHash function ?

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [reply packages/protocol/contracts/governance/Governance.sol:664]: On second thought this will work because of onlyApproverOrSecurityCouncil. Can we refactor this so it is more readable ?

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol]: added additional condition

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol:664]: What would you suggest in terms of readability improvement?

Comment (pahor167) [reply packages/protocol/contracts/governance/Governance.sol:664]: Probably removing modifier since we are checking approver twice and keep the logic only in this function.

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol]: I explained [here](https://github.com/celo-org/celo-monorepo/pull/11014#issuecomment-2146175966). But the gist is that it prevents a hotfix from being executed long after it was approved and requires re-approval if too much time elapsed since it was last prepared.

Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol]: I extracted it to an public function and added tests, as of [fbba0ea](https://github.com/celo-org/celo-monorepo/pull/11014/commits/fbba0ea7f3a81f385152354c00acc58c50d5e54d)

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol:664]: Not sure it improved readability, but I removed the modifier and included the logic in the function body.

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol:729]: I would rather keep it as is, to avoid having to modify the `getHotfixHash` to make it public, so as not to call `this.getHotfixHash`

Comment (pahor167) [reply packages/protocol/contracts/governance/Governance.sol:729]: Do you have any concern with making getHotfixHash public ?

Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol:729]: It just ends up costing more gas if I use the function.

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol:228]:
```suggestion
    require(hotfixes[hash].executionTimeLimit > 0, "hotfix time limit not set");
```

File: packages/protocol/contracts/governance/Governance.sol
Comment (pahor167) [inline packages/protocol/contracts/governance/Governance.sol:740]: Shouldn't this be rather executionTimeLimit < now since when executionTimeLimit >= now it is considered as timed out and can be reset ?

Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol:740]: `executionTimeLimit >= now` is still within the time `hotfixExecutionTimeWindow`. To reset the hotfix it needs to wait the the execution window passed. 

```require(hotfixes[hash].executionTimeLimit < now, "hotfix execution time limit not reached");```

check out the `GovernanceTest_resetHotfix` for the intended execution flow.

**Review by pahor167 — APPROVED** [review]

**Review by pahor167 — APPROVED** [review]

File: packages/protocol/contracts/governance/Governance.sol
Comment (aaronmgdr) [inline packages/protocol/contracts/governance/Governance.sol:669]: would we want ta different event for security council approval vs regular approval?

File: packages/protocol/test-sol/unit/governance/network/Governance.t.sol
Comment (aaronmgdr) [inline packages/protocol/test-sol/unit/governance/network/Governance.t.sol]: this test could have. false postive i believe if the approveHotfix fails it will never be true. could add an intermediate assertion

File: packages/protocol/contracts/governance/Governance.sol
Comment (soloseng) [reply packages/protocol/contracts/governance/Governance.sol:669]: We could have a different one, but I felt that having the different address included in the event would take care of differenciating the two.

**Pull request merged**
